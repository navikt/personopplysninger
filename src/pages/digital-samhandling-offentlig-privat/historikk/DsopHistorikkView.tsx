import { Fragment, useState } from 'react';
import { DsopInfo } from 'types/dsop';
import { FormattedMessage } from 'react-intl';
import { Link as ReactLink, useLocation } from 'react-router-dom';
import { Alert, Label, Link } from '@navikt/ds-react';
import { ChevronUpIcon, ChevronDownIcon } from '@navikt/aksel-icons';
import dayjs from 'dayjs';

interface Props {
    dsopInfo: DsopInfo;
}

export const DsopHistorikkView = (props: Props) => {
    const location = useLocation();
    const { dsopInfo } = props;

    const initState: {
        [year: string]: {
            innslag: DsopInfo;
            ekspandert: boolean;
        };
    } = {};

    dsopInfo.forEach((innslag, i) => {
        const year = dayjs(innslag.uthentingsTidspunkt).year();

        if (!initState[year]) {
            initState[year] = {
                innslag: [innslag],
                ekspandert: !i,
            };
        } else {
            initState[year].innslag.push(innslag);
        }
    });

    const [data, setData] = useState(initState);

    return (
        <div className="historikk__tabs-innhold historikk__flex-table">
            {Object.keys(data).length > 0 ? (
                <>
                    <div className={'historikk__info'}>
                        <Alert variant="info">
                            <FormattedMessage
                                id={'lenker.dsop.info'}
                                values={{
                                    a: (text) => (
                                        <Link href="https://altinn.no" target="blank">
                                            {text}
                                        </Link>
                                    ),
                                }}
                            />
                        </Alert>
                    </div>
                    <div className="historikk__flex-rad historikk__head">
                        <div className="historikk__flex-kolonne">
                            <Label as="p">
                                <FormattedMessage id="dsop.uthentingstidspunkt" />
                            </Label>
                        </div>
                        <div className="historikk__flex-kolonne">
                            <Label as="p">
                                <FormattedMessage id="dsop.mottaker" />
                            </Label>
                        </div>
                    </div>
                    {Object.keys(data)
                        .reverse()
                        .map((year) => {
                            const value = data[year];

                            const onClick = () => {
                                setData({
                                    ...data,
                                    [year]: {
                                        ...data[year],
                                        ekspandert: !data[year].ekspandert,
                                    },
                                });
                            };

                            return (
                                <Fragment key={year}>
                                    <div className="historikk__flex-rad" key={year}>
                                        <button className="historikk__flex-kolonne af-liste__ekspander" onClick={onClick}>
                                            {year} {value.ekspandert ? <ChevronUpIcon aria-hidden="true" /> : <ChevronDownIcon aria-hidden="true" />}
                                        </button>
                                        <div />
                                    </div>
                                    {value.ekspandert &&
                                        value.innslag.map((innslag, i) => (
                                            <div className="historikk__flex-rad" key={i}>
                                                <div className="historikk__flex-kolonne historikk__heading">
                                                    {dayjs(innslag.uthentingsTidspunkt).format('DD.MM kl. hh:mm')}
                                                </div>
                                                <div className="historikk__flex-kolonne">
                                                    <Link as={ReactLink} to={`${location.pathname}/${innslag.uthentingsTidspunkt}`} className="lenke">
                                                        {innslag.mottakernavn}
                                                    </Link>
                                                </div>
                                            </div>
                                        ))}
                                </Fragment>
                            );
                        })}
                </>
            ) : (
                <div className="historikk__ingen-data">
                    <Alert variant="info">
                        <FormattedMessage id="dsop.ingendata" />
                    </Alert>
                </div>
            )}
        </div>
    );
};

export default DsopHistorikkView;
