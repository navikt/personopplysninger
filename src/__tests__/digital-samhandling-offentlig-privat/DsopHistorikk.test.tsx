import { IntlProvider } from 'react-intl';
import { render } from '@testing-library/react';
import dsopInfo from '@/clients/apiMock/app/fetch/dsop-info.json';
import nbMessages from '@/text/nb';
import DsopHistorikkView from '@/pages/digital-samhandling-offentlig-privat/historikk/DsopHistorikkView';

vi.mock('react-router-dom', () => ({
    Link: () => <div />,
    useLocation: () => ({
        pathname: 'pathname',
    }),
}));

describe('DsopHistorikkView', () => {
    it('renders correctly', () => {
        const { asFragment } = render(
            <IntlProvider locale={'nb'} messages={nbMessages}>
                <DsopHistorikkView dsopInfo={dsopInfo} />
            </IntlProvider>
        );
        expect(asFragment()).toMatchSnapshot();
    });
});
