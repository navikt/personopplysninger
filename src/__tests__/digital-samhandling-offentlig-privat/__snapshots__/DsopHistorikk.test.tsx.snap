// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DsopHistorikkView renders correctly 1`] = `
<DocumentFragment>
  <div
    class="historikk__tabs-innhold historikk__flex-table"
  >
    <div
      class="historikk__info"
    >
      <div
        class="navds-alert navds-alert--info navds-alert--medium"
      >
        <svg
          aria-labelledby="title-r0"
          class="navds-alert__icon"
          fill="none"
          focusable="false"
          height="1em"
          role="img"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <title
            id="title-r0"
          >
            Informasjon
          </title>
          <path
            clip-rule="evenodd"
            d="M0 12C0 5.382 5.382 0 12 0c6.617 0 12 5.382 12 12s-5.383 12-12 12C5.382 24 0 18.618 0 12Zm12-7a1.5 1.5 0 1 1 0 3 1.5 1.5 0 0 1 0-3ZM9 19v-2h2v-5H9v-2h4v7h2v2H9Z"
            fill="#368DA8"
            fill-rule="evenodd"
          />
        </svg>
        <div
          class="navds-alert__wrapper navds-body-long"
        >
          Dette er en logg som viser hvilke opplysninger vi har utlevert til forsikringsselskap basert på samtykke du har gitt på Altinn. 
          <a
            class="navds-link"
            href="https://altinn.no"
            target="blank"
          >
            Se samtykket i Altinn.
          </a>
        </div>
      </div>
    </div>
    <div
      class="historikk__flex-rad historikk__head"
    >
      <div
        class="historikk__flex-kolonne"
      >
        <p
          class="navds-label"
        >
          Uthentingstidspunkt
        </p>
      </div>
      <div
        class="historikk__flex-kolonne"
      >
        <p
          class="navds-label"
        >
          Selskap
        </p>
      </div>
    </div>
    <div
      class="historikk__flex-rad"
    >
      <div
        class="historikk__flex-kolonne af-liste__ekspander"
      >
        2019 
        <svg
          fill="none"
          focusable="false"
          height="1em"
          role="img"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="m12 6.586 7.707 7.707-1.414 1.414L12 9.414l-6.293 6.293-1.414-1.414L12 6.586Z"
            fill="currentColor"
            fill-rule="evenodd"
          />
        </svg>
      </div>
      <div />
    </div>
    <div
      class="historikk__flex-rad"
    >
      <div
        class="historikk__flex-kolonne historikk__heading"
      >
        18.06 kl. 10:45
      </div>
      <div
        class="historikk__flex-kolonne"
      >
        <div />
      </div>
    </div>
    <div
      class="historikk__flex-rad"
    >
      <div
        class="historikk__flex-kolonne historikk__heading"
      >
        18.06 kl. 10:45
      </div>
      <div
        class="historikk__flex-kolonne"
      >
        <div />
      </div>
    </div>
    <div
      class="historikk__flex-rad"
    >
      <div
        class="historikk__flex-kolonne historikk__heading"
      >
        18.06 kl. 10:51
      </div>
      <div
        class="historikk__flex-kolonne"
      >
        <div />
      </div>
    </div>
    <div
      class="historikk__flex-rad"
    >
      <div
        class="historikk__flex-kolonne af-liste__ekspander"
      >
        2018 
        <svg
          fill="none"
          focusable="false"
          height="1em"
          role="img"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M12 17.414 4.293 9.707l1.414-1.414L12 14.586l6.293-6.293 1.414 1.414L12 17.414Z"
            fill="currentColor"
            fill-rule="evenodd"
          />
        </svg>
      </div>
      <div />
    </div>
    <div
      class="historikk__flex-rad"
    >
      <div
        class="historikk__flex-kolonne af-liste__ekspander"
      >
        2017 
        <svg
          fill="none"
          focusable="false"
          height="1em"
          role="img"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M12 17.414 4.293 9.707l1.414-1.414L12 14.586l6.293-6.293 1.414 1.414L12 17.414Z"
            fill="currentColor"
            fill-rule="evenodd"
          />
        </svg>
      </div>
      <div />
    </div>
    <div
      class="historikk__flex-rad"
    >
      <div
        class="historikk__flex-kolonne af-liste__ekspander"
      >
        2016 
        <svg
          fill="none"
          focusable="false"
          height="1em"
          role="img"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M12 17.414 4.293 9.707l1.414-1.414L12 14.586l6.293-6.293 1.414 1.414L12 17.414Z"
            fill="currentColor"
            fill-rule="evenodd"
          />
        </svg>
      </div>
      <div />
    </div>
  </div>
</DocumentFragment>
`;
